### PR TITLE
Add Spring framework integration

### LINKED ISSUE #61: Spring Boot integration
We did not open-source our Spring Boot integration yet, but we wish to soon.

### PR DESCRIPTION
**Pull request by trustin**
Note: Special thanks to @imasahiro who is the initial author of Spring
framework integration.

Modifications:

- Add client-java-spring project
- Update NOTICE.txt for Spring framework and Hibernate Validator
- Add documentation for Spring integration

Result:

- Fixes #61

### CODE REVIEW

File: client/java-spring/src/main/java/com/linecorp/centraldogma/client/spring/CentralDogmaClientSettings.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    /**
+     * Returns the member endpoints of this Central Dogma endpoint group.
+     */
+    public List<String> endpoints() {
+        return endpoints;
+    }
+
+    /**
+     * Sets the member endpoints of this Central Dogma endpoint group.
```
Comment (imasahiro) [inline client/java-spring/src/main/java/com/linecorp/centraldogma/client/spring/CentralDogmaClientSettings.java]: How about specify endpoint format as well?
```
Each endpoints may have one of the follwing formats:
* hostname
* hostname:port
```

File: settings.gradle [hunk 2]
```diff
@@ -2,6 +2,7 @@ rootProject.name = 'centraldogma'
 
 // Published Java projects
 include 'client:java'
+include 'client:java-spring'
```
Comment (imasahiro) [inline settings.gradle]: Think it is better to follow spring-boot module construction manner.
How about renaming this module to client:java-spring-boot-autoconfigure and add another module named client:java-spring-boot-starter that only depends on autoconfigure module?

Comment (trustin) [reply settings.gradle]: Renamed to `centraldogma-client-spring-boot-autoconfigure` and added `centraldogma-client-spring-boot-starter`.

File: client/java-spring/src/main/java/com/linecorp/centraldogma/client/spring/CentralDogmaClientSettings.java (hunk 1 above)
Comment (trustin) [reply client/java-spring/src/main/java/com/linecorp/centraldogma/client/spring/CentralDogmaClientSettings.java]: Fixed

File: client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogmaBuilder.java [hunk 3]
```diff
@@ -29,30 +45,195 @@
...
      * @param uri the URI of the Central Dogma server. e.g.
      *            {@code tbinary+http://example.com:36462/cd/thrift/v1}
      */
+    @Deprecated
     public CentralDogmaBuilder uri(String uri) {
-        this.uri = URI.create(requireNonNull(uri, "uri"));
+        final URI parsed = URI.create(requireNonNull(uri, "uri"));
+        final String host = parsed.getHost();
+        final int port = parsed.getPort();
+        checkArgument(host != null, "uri: %s (must contain a host part)");
```
Comment (imasahiro) [inline client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogmaBuilder.java]: Could you add `host` to arguments to show up `host` to error message?
`checkArgument(host != null, "uri: %s (must contain a host part)", host)`

File: client/java-spring-boot-autoconfigure/build.gradle [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Centraloga client Spring(-boot) integration
```
Comment (minwoox) [inline client/java-spring-boot-autoconfigure/build.gradle]: Centraloga?

File: client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java [hunk 5]
```diff
@@ -42,21 +42,120 @@
 public interface CentralDogma {
 
     /**
+     * Returns a newly-created {@link CentralDogma} instance which connects to the given host at port 36462
+     * using the the default {@link ClientFactory}.
```
Comment (minwoox) [inline client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java]: the the?

File: client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java [hunk 6]
```diff
@@ -42,21 +42,120 @@
...
+     *
+     * @param host the host name or IP address of the Central Dogma server
+     */
+    static CentralDogma forHost(String host) {
+        return new CentralDogmaBuilder().host(host).build();
+    }
+
+    /**
+     * Returns a newly-created {@link CentralDogma} instance which connects to the given host and port
+     * using the the default {@link ClientFactory}.
```
Comment (minwoox) [inline client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java]: the the?

File: client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java [hunk 7]
```diff
@@ -42,21 +42,120 @@
...
+     * @param host the host name or IP address of the Central Dogma server
+     * @param port the port number of the Central Dogma server
+     */
+    static CentralDogma forHost(String host, int port) {
+        return new CentralDogmaBuilder().host(host, port).build();
+    }
+
+    /**
+     * Returns a newly-created {@link CentralDogma} instance which connects to the given host at port 36462
+     * using the the specified {@link ClientFactory}.
```
Comment (minwoox) [inline client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java]: the the

File: client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java [hunk 8]
```diff
@@ -42,21 +42,120 @@
...
+     */
+    static CentralDogma forHost(ClientFactory clientFactory, String host) {
+        return new CentralDogmaBuilder().clientFactory(clientFactory)
+                                        .host(host)
+                                        .build();
+    }
+
+    /**
+     * Returns a newly-created {@link CentralDogma} instance which connects to the given host and port
+     * using the the specified {@link ClientFactory}.
```
Comment (minwoox) [inline client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogma.java]: the the

File: client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogmaBuilder.java (hunk 3 above)
Comment (trustin) [reply client/java/src/main/java/com/linecorp/centraldogma/client/CentralDogmaBuilder.java]: Oops. Added `uri` as the last parameter of `checkArgument()`

**Comment by minwoox** [comment]
LGTM~ 👍

**Pull request merged**